File: src/core/validators/SymbolTable.cs
using System.Collections.Generic;
using System.Linq;
using Antlr4.Runtime;
using MetaEd.Grammar.Antlr;

namespace MetaEd.Core.Validator
{
    public interface ISymbolTable
    {
        bool TryAdd(string entityType, string name, ParserRuleContext context);
        EntityContext Get(string entityType, string name);
        bool IdentifierExists(string entityType, string identifier);
        IEnumerable<string> IdentifiersForEntityType(string entityType);
        IEnumerable<string> IdentifiersForEntityProperties(string entityType, string name);
        IEnumerable<IPropertyWithComponents> ContextsForMatchingPropertyIdentifiers(string entityType, string name, ICollection<string> candidatePropertyIdentifiers);
    }

    public class SymbolTable : ISymbolTable
    {
        private readonly Dictionary<string, Dictionary<string, EntityContext>> _symbolTable = new Dictionary<string, Dictionary<string, EntityContext>>();

        public bool TryAdd(string entityType, string name, ParserRuleContext context)
        {
            Dictionary<string, EntityContext> entityDictionary;

            if (!_symbolTable.TryGetValue(entityType, out entityDictionary))
            {
                entityDictionary = new Dictionary<string, EntityContext>();
                _symbolTable.Add(entityType, entityDictionary);
            }

            if (entityDictionary.ContainsKey(name))
                return false;

            var entityContext = new EntityContext
                                    {
                                        Name = name,
                                        Context = context
                                    };
            entityContext.PropertySymbolTable = new PropertySymbolTable(entityContext);
            entityDictionary.Add(name, entityContext);
            return true;
        }

        public EntityContext Get(string entityType, string name)
        {
            Dictionary<string, EntityContext> entityDictionary;
            if (!_symbolTable.TryGetValue(entityType, out entityDictionary))
                return null;

            EntityContext entityContext = null;
            entityDictionary.TryGetValue(name, out entityContext);
            return entityContext;
        }

        public bool IdentifierExists(string entityType, string identifier)
        {
            if (!_symbolTable.ContainsKey(entityType)) return false;
            return _symbolTable[entityType].ContainsKey(identifier);
        }

        public IEnumerable<string> IdentifiersForEntityType(string entityType)
        {
            Dictionary<string, EntityContext> entityDictionary;
            return _symbolTable.TryGetValue(entityType, out entityDictionary) ? entityDictionary.Keys.AsEnumerable() : Enumerable.Empty<string>();
        }

        // results are prefixed by a 'with context' value if one exists for property
        public IEnumerable<string> IdentifiersForEntityProperties(string entityType, string identifier)
        {
            var entityContext = Get(entityType, identifier);

            if (entityContext == null) return Enumerable.Empty<string>();
            return entityContext.PropertySymbolTable.Identifiers();
        }

        // candidate identifiers should be prefixed by a 'with context' value if one exists for property
        public IEnumerable<IPropertyWithComponents> ContextsForMatchingPropertyIdentifiers(string entityType, string identifier, ICollection<string> candidatePropertyIdentifiers)
        {
            var entityContext = Get(entityType, identifier);

            if (entityContext == null) return Enumerable.Empty<IPropertyWithComponents>();
            return entityContext.PropertySymbolTable.ContextsForMatchingIdentifiers(candidatePropertyIdentifiers);
        }
    }

    public class EntityContext
    {
        public string Name { get; set; }
        public ParserRuleContext Context { get; set; }
        public PropertySymbolTable PropertySymbolTable { get; set; }
    }
}

File: src/core/validators/SymbolTable.ts
import {IPropertyWithComponents} from '../../grammar/IPropertyWithComponents';
import PropertySymbolTable from './PropertySymbolTable'

export interface ISymbolTable {
    tryAdd(entityType: string, name: string, context: any): boolean;
    get(entityType: string, name: string): EntityContext;
    identifierExists(entityType: string, identifier: string): boolean;
    identifiersForEntityType(entityType: string): IterableIterator<string>;
    identifiersForEntityProperties(entityType: string, identifier: string): IterableIterator<string>;
    contextsForMatchingPropertyIdentifiers(entityType: string, name: string, candidatePropertyIdentifiers: Array<string>): Array<any>;
}

export default class SymbolTable implements ISymbolTable {

    private symbolTable: Map<string, Map<string, EntityContext>> = new Map<string, Map<string, EntityContext>>();

    public tryAdd(entityType: string, name: string, context: any): boolean {

        let entityDictionary: Map<string, EntityContext> = this.symbolTable.get(entityType);

        if (!entityDictionary) {
            entityDictionary = new Map<string, EntityContext>();
            this.symbolTable.set(entityType, entityDictionary);
        }

        if (entityDictionary.has(name))
            return false;

        let entityContext = new EntityContext(name, context);

        entityContext.propertySymbolTable = new PropertySymbolTable(entityContext);
        entityDictionary.set(name, entityContext);

        return true;
    }

    public get(entityType: string, name: string): EntityContext {
        let entityDictionary: Map<string, EntityContext> = this.symbolTable.get(entityType);
        if (!entityDictionary)
            return null;

        let entityContext: EntityContext = entityDictionary.get(name);

        return entityContext;
    }

    public identifierExists(entityType: string, identifier: string): boolean {
        if (!this.symbolTable.has(entityType)) return false;
        return this.symbolTable[entityType].containsKey(identifier);
    }

    public identifiersForEntityType(entityType: string): IterableIterator<string> {
        let entityDictionary: Map<string, EntityContext> = this.symbolTable.get(entityType);
        if (entityDictionary) return entityDictionary.keys();
        return new Array<string>().values();
    }

    // results are prefixed by a 'with context' value if one exists for property
    public identifiersForEntityProperties(entityType: string, identifier: string): IterableIterator<string> {
        let entityContext = this.get(entityType, identifier);

        if (entityContext == null)
            return [];
        return entityContext.propertySymbolTable.identifiers();
    }

    // candidate identifiers should be prefixed by a 'with context' value if one exists for property
    public contextsForMatchingPropertyIdentifiers(entityType: string, name: string, candidatePropertyIdentifiers: Array<string>): Array<IPropertyWithComponents> {
        let entityContext = this.get(entityType, name);

        if (entityContext == null) return [];
        return entityContext.propertySymbolTable.contextsForMatchingIdentifiers(candidatePropertyIdentifiers);
    }
}

export class EntityContext {
    public propertySymbolTable: PropertySymbolTable;
    public name: string;
    public context: any;
    public constructor (name: string, context: any) {
        this.name = name;
        this.context = context;
    }
}
